raft.go

```diff
@@ -665,6 +790,54 @@ func (r *Raft) verifyLeader(v *verifyFuture) {
 	}
 }
 
+// leadershipTransfer is doing the heavy lifting for the leadership transfer.
+func (r *Raft) leadershipTransfer(id ServerID, address ServerAddress, repl *followerReplication, stopCh chan struct{}, doneCh chan error) {
+
+	// make sure we are not already stopped
+	select {
+	case <-stopCh:
+		doneCh <- nil
+		return
+	default:
+	}
+
+	// Step 1: set this field which stops this leader from responding to any client requests.
+	r.setLeadershipTransferInProgress(true)
+	defer func() { r.setLeadershipTransferInProgress(false) }()
+
+	for repl.nextIndex <= r.getLastIndex() {
```

Comment:
This access of the `followerReplication.nextIndex` made me wonder about thread safety. Presumably this is racey since the replication thread is still running and could be writing to the `nextIndex` concurrently?

Reply:
Using `atomic` now to protect from the race. Great catch!